### ISSUE TITLE
Provide entry point, step-based guide and better UX for end users

### ISSUE THREAD

**Issue by ocram** [urbandroid-team/dont-kill-my-app#34]
As Android developers, we have all had to fight against these issues one time or another. Now, there’s finally some project that curates resources related to this problem. Thanks a lot for this!

The current website (`dontkillmyapp.com`) is already perfect for developers. But for end users, the UX could be improved quite a bit.

We have all [needed](https://github.com/urbandroid-team/dont-kill-my-app/issues/16) to give users step-by-step advice on how to make our app work on their devices again. We all need some link that we can share with our users to guide them on what to do. But the current website is not really useful for that. Here’s what would be necessary (in my opinion) to make this an awesome resource that can be shared with users (e.g. in customer support):

 1. The lead copy (“We have to fight back!”) is targeted towards developers rather than end users. If the website accepted `app-name` as a parameter in the query string, you could turn the lead copy into something like this (if the parameter is not empty): “Due to some modifications made to your operating system by your device vendor, background tasks don’t work well for you with {app-name}. Here are some simple steps showing how you can easily fix this.”
 1. After the vendor selection (which is perfect), the vendor’s details page should not include the “Story” for end users. It’s great for developers, but end users are not interested in this. They just want their problem solved and forget about this.
 1. The “Solution for developers” should not appear on the vendor’s details page for end users, obviously.
 1. “Solution for users” should be the only content on the vendor’s details page. But its different device-specific and version-specific pieces of advice should be hidden behind another level of navigation, one that asks for the user’s device.
 1. Users generally don’t know their Android version. We have to ask for their device name. If you care about the whole market and want your app to succeed, you need to address the 85% of non-technical users for whom their device’s name is the only thing they know about their device, or who have never heard of “Android” – just “Nokia”.
 1. Instead of displaying an item for [“Stock Android”](https://github.com/urbandroid-team/dont-kill-my-app/issues/24), we’d have to show separate items for popular devices with stock Android, such as the Pixel devices. Yes, 85% of users don’t know their Pixel device is what we call stock Android.
 1. [Screenshots](https://bitbucket.org/copluk/acr/issues/607) are something that end users will appreciate a lot. If every step is accompanied by a screenshot, which is something that can definitely be done by us in the developer community, they will easily follow all steps and be more satisfied with our apps again.
 1. The `/?3` version should probably be the default for the link geared towards the needs of end users. But that wouldn’t necessarily require changes in the code. Developers could generate their links as `/?3&app-name=My+app`.
 1. After all, [internationalization](https://github.com/urbandroid-team/dont-kill-my-app/issues/19) would help a lot as well, of course.

Naturally, I would be willing to contribute if the determination to do this and the required infrastructure was there. Happy users means more success for our apps.

Thanks!

**Comment by yccheok** [comment]
Agree. This site might be the key, to raise end users awareness. Nothing will change, if only developers aware on this issue. 

AlarmManager will continue broken.

However, if end users aware of this, that can pressure Google to do something about it. At least enforce a unify battery optimization policy, or at least make AlarmManager not impacted by battery optimization policy.

**Comment by Artaud** [comment]
Many thanks for the great points. I agree with most of them.

I'll tie points 1,2,3,8 to the _app_ URL parameter.

Not sure about point 4. With most of vendors, we currently don't have specific steps for specific devices, so maybe we might think about this later when the pages are more cluttered?

5) I believe this is not an issue as the main navigation lists device names?

6) We renamed the Stock Android to Google (Pixel, Nexus) - what do you think about that?

7) Agree completely. The only issue is getting the screenshots. This is something we'd need from the developer community as we don't have access to most of the phones to generate the screenshots.

9) No clue how to do that. Technically, this is easy, but keeping it up to date sounds like an impossible thing, and frankly, in the age of Google Translate, as imperfect as it may be, I think we're still never going to beat that...

Let me know your further thoughts....

**Comment by Artaud** [comment]
@yccheok 
> However, if end users aware of this, that can pressure Google to do something about it. At least enforce a unify battery optimization policy, or at least make AlarmManager not impacted by battery optimization policy.

This is a great point and ideally the ultimate goal of this site. If Google would be able to keep the battery optimization under control, the battle is won.

**Comment by ocram** [comment]
Thanks for your feedback!

> I'll tie points 1,2,3,8 to the `app` URL parameter.

Yes, that would be perfect.

> Not sure about point 4. With most of vendors, we currently don't have specific steps for specific devices, so maybe we might think about this later when the pages are more cluttered?

Of course, you don’t have specific steps for every device. But the problem is that if you name it just “Samsung”, users with an older S6 will visit the page and try the steps that may actually be for the latest S9. So there has to be *some* differentiation. Perhaps just “Android version” as a next step. But with another step in the navigation, the Android version would still work better than listing all the different Android versions below each other in a long wall of text (and some screenshots).

> I believe this is not an issue as the main navigation lists device names?

Sure, where device names are available, that’s always better than using an Android version name in communication towards end users. Sometimes, we may have to distinguish between “S6 (Android 5)” and “S6 (Android 7.1)” though where Android updates have been introduced with major changes.

> We renamed the Stock Android to Google (Pixel, Nexus) - what do you think about that?

Yes, that’s great. Although where Motorola or other vendors with near-stock Android have the same issues, we may have to use an “alias” that has the same content as the Google category.

> Agree completely. The only issue is getting the screenshots. This is something we'd need from the developer community as we don't have access to most of the phones to generate the screenshots.

Absolutely. As I said, I would be willing to contribute, and others will surely be, too.

> No clue how to do that. Technically, this is easy, but keeping it up to date sounds like an impossible thing, and frankly, in the age of Google Translate, as imperfect as it may be, I think we're still never going to beat that...

Well, as I said, translation is definitely nice to have, but certainly not as critical a factor as the other issues. We should definitely postpone this, at least.

**Comment by Artaud** [comment]
I've pushed the ?app URL parameter in a1d6499f8d3560ff6d5dc458fd695e6770cdcb08

You can now use params like dontkillmyapp.com/nokia?3&app=JStock and all occurences of "your app" will be replaced by JStock, and only the relevant information for end users will be shown.

**Comment by ocram** [comment]
Thanks a lot!

This is awesome, and solves most of what this issue is about. The addition of screenshots (or at least some infrastructure for that, e.g. guidance on where to add them) would be something that’s left.

What `makeSiteUserFacing` does is fine, though it’s not really sufficient to apply the text replacement to `p` only. It should probably be `#user-solution-section` instead, as the text “your app” does also occur in `ul`, etc.

Moreover, in `#user-solution-title`, you could probably replace `Solution for users` with just `Solution` as it’s the only thing left if it’s a user-facing page.

**Comment by Artaud** [comment]
yep I've noticed, I'll push a fix in a few moments

**Comment by Artaud** [comment]
fixed with 0ebb6aac6fb6e061130ada308c6769747b83d8c0

**Comment by ocram** [comment]
Thank you!

The last thing I noticed is that you cannot open `/?3&app=JStock` to have this set on the overview page. At least, the query string doesn’t get appended to the individual vendor links, which works on the vendor page, though.

Finally, I guess, we’d just need a single short paragraph of documentation somewhere to tell developers that this is possible and that they can refer their users to the pages with a custom link.

**Comment by Artaud** [comment]
Docs have arrived in the API section, including also a Java example on using the JSON API in a webview in an app for quick usage

**Comment by matejdro** [comment]
Maybe adb-based solutions should be grouped up under Advanced title or maybe even hidden by default?  Most users do not know how to do these and it might overwhelm them or turn them away from even doing basic solution above.

**Comment by ocram** [comment]
@Artaud Thanks a lot for addressing many of the points outlined in the original post!

Still, I would say your current website (`dontkillmyapp.com`) is *good* for developers, *perfect* for drawing attention to this problem and building public awareness, underlining the need for change, but it’s hardly perfect for end users.

Now maybe that is what you want. Attracting attention from the press or even vendors is great. We definitely need that for this problem.

But supporting users with these problems, and thus improving user satisfaction, is important as well. Throughout the last month, we have been making public [the solution that we have been using](http://www.undoze.com/) as well. Our solutions seem to serve different purposes, but in any case, they are complementary.

I’m sure multiple solutions can benefit from each other and from an exchange of ideas. Or perhaps you’re even willing to further improve this project for end users, which would be great.

**Comment by otede** [comment]
It's amazing how oblivious intelligent people can be. 
1. Open ReVanced patched youtube
2. GmsCore is not whitelisted from battery optimization. Follow the "Don't kill my app" GUIDE for GmsCore. [OPEN WEBSITE]
3. https://dontkillmyapp.com/
4. there
5. is
6. no guide
